keps/sig-api-machinery/3610-env-injection/README.md

```diff
@@ -0,0 +1,855 @@
+
+- for cluster-wide, name of the annotaiton or configmap can be `cluster-wide-envs`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A global setting like proxy or env that used by default in some clients.
+
+- This would also be very useful when running a service mesh like Linkerd within the cluster, as you have to set the http_proxy environment for everything that should use it.
+
+
+#### Story 2
+
+From <https://github.com/kubernetes/kubernetes/issues/48180#issuecomment-331941326>
+> Automatically inject the kubernetes namespace into an environment variable for me. With dozens of different kube templates people sometimes
+> forget to expose it manually (<https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/>)
+> Have compute nodes have special knowledge about themselves that can be injected into environment variables. We have an odd/even compute node
+> concept for upgrading openshift and this information may come in handy
+> Be able to specify system-wide environment variables. We have multiple openshift instances (different environments and different data centers).
+> It would be great if that data could be made automatically available without having to deal with custom env sections in configs.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+Annotation in namespace cannot fit the cluster-wide case.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+As there is no validation for the key value in annotation or configmap, users will get errors util they create a new pod.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+1. If a pod has already set the env, skip it. Overwriting is too aggressive.
+If users want to overwrite the env, `global.env/overwrite: true` should be set in namespace annotation.
+
+2. Env injection only for pod creation. Not for update(The behavior that a pod update may change its env, is odd).
```

Comment:
for pod level, do you mean adding env to every container and init container of the pod?

Reply:
say I create the namespace with ENV `x=y,a=b`. Over a period of time, I change `x=z,a=b` How will this value be sent to the pod since the KEP says we won't support the ENV change in POD update case, which is perfectly valid. So, someone will have to delete all the pods in suitable namespace and let them be re-created with new ENV ? case in point is something like proxy with password where the password might changes over a period of time.

Reply:
I think this depens on `Should we overwriting existed env? Or do need we to add a flag`

Pod is always created by a controller like deployment controller or job controller. If the deployment is without the env, new pods that are created from the deployment would have new env from new injection settings. I think overwriting is too aggressive by default. However, this may be a use case and users may want to overwrite for some reason.

If overwrite is an always behavior, pod update can be supported. Hence, pod env can be overwritten to the latest settings once updated.

Reply:
True. That begs the question, if we are to do these kind of injection, won't it be better off done at the `Deployment/STS/Job/CronJob` level than the eventual `Pod` level ? Since trying to patch a running pod on a namespace to inject or update and env will lead to an error `error: failed to patch env update to pod template:....*` anyway. One has to delete the pod and let it recreate with new Env. But if this has to be dealt with at the pod level then we kind of end up bypassing all the good bigs provided by deployments/sts other in terms of their rolling update nature.  

I am partial to saying, if the global env on the namespace is updated, then users will have to explicitly perform a rollout of the resources in that namespace so that the pods of the said resources can inherit the new env ?

Reply:
> I think overwriting is too aggressive by default.

This I agree on. The default should always be ignoring the overwrite. If the env exists, it should have been added as a conscious decision on the end user part. 

But what I was mainly concerned about was the change of env that was injected to the Pod spec by the admission controller. That also means, we would have to have a way to distinguish that if the admission controller did add the ENV or it opted not to because there was an override at the  Pod spec level already. We have done something similar by just adding an annotation saying if the ENV was indeed injected by the Admission controller or not. If it was not injected by the admission controller, all possible further actions will ignore such pods honouring the default override nature.

Reply:
> This I agree on. The default should always be ignoring the overwrite. If the env exists, it should have been added as a conscious decision on the end user part.

I changed to using API Object instead of annotation. I followed the `LimitRange` design and the default will be applied if it is not set. If we want to overwrite, I add another `forceEnv` like `hard`(default like `soft`).